File: console/Mpro.vb
Imports System.Net.Sockets
Imports K2Informatics.Erlnet

Public Class Mpro
    Inherits Imem

    Protected Sub New(ByRef _stream As NetworkStream)
        MyBase.New(_stream)
    End Sub

    Public Overloads Shared Function Connect(ByVal host As String, ByVal port As Integer) As Mpro
        Dim client As TcpClient = New TcpClient(host, port)
        Dim s As NetworkStream = client.GetStream()
        Return New Mpro(s)
    End Function

    Public Function listDestinationChannels(ByVal AppId) As ArrayList
        Return CallMproMFASync("listDestinationChannels", AppId)
    End Function
    Public Function putSourcePeerName(ByRef AppId As String, ByRef Key As String, ByRef SpName As String) As ArrayList
        Return CallMproMFASync("putSourcePeerName", AppId, Key, SpName)
    End Function
    Public Function deleteSourcePeer(ByRef AppId As String, ByRef Key As String) As ArrayList

        Return CallMproMFASync("deleteSourcePeer", AppId, Key)
    End Function
    Public Function listSourcePeerKeys(ByRef AppId As String) As ArrayList
        Return CallMproMFASync("listSourcePeerKeys", AppId)
    End Function
    Public Function listSourcePeers(ByRef AppId As String) As ArrayList
        Return CallMproMFASync("listSourcePeers", AppId)
    End Function
    Public Function putSourcePeer(ByRef AppId As String, ByRef Key As String, ByRef ChStr As String, ByRef OptStr As String) As ArrayList
        Return CallMproMFASync("putSourcePeer", AppId, Key, ChStr, OptStr)
    End Function
    Public Function putWhitelist(ByRef AppId As String, ByRef Key As String, ByRef IpAddressStr As String, ByRef OptStr As String) As ArrayList
        Return CallMproMFASync("putWhitelist", AppId, Key, IpAddressStr, OptStr)
    End Function
    Public Function deleteWhitelist(ByRef AppId As String, ByRef Key As String, ByRef IpAddressStr As String) As ArrayList
        Return CallMproMFASync("deleteWhitelist", AppId, Key, IpAddressStr)
    End Function
    Public Function getSourcePeer(ByRef AppId As String, ByRef Key As String) As ArrayList
        Return CallMproMFASync("getSourcePeer", AppId, Key)
    End Function
    Public Function getWhitelist(ByRef AppId As String, ByRef Key As String) As ArrayList
        Return CallMproMFASync("getWhitelist", AppId, Key)
    End Function
    Private Function CallMproMFASync(ByVal fun As String, ByVal ParamArray argsRest() As Object) As ArrayList
        Return CallImemMFASync("mpro_dal_prov", fun, argsRest)
    End Function

End Class

File: lib/Mpro.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Net.Sockets;
using System.Collections;
using K2Informatics.Erlnet;

namespace K2Informatics.Erlimemnet
{
    public class Mpro : Imem
    {
        public Mpro(ErlStream _stream) : base(_stream) { }

        public static Mpro Connect(ref string host, int port, bool secure)
        {
            return (Mpro)Imem.Connect(ref host, port, secure, typeof(Mpro));
        }

        public object[] listDestinationChannels(ref string AppId)
        {
            return CallMproMFASync("listDestinationChannels", AppId);
        }

        public object[] deleteSourcePeer(ref string AppId, ref string Key)
        {
            return CallMproMFASync("deleteSourcePeer", AppId, Key);
        }

        public object[] listSourcePeerKeys(ref string AppId)
        {
            return CallMproMFASync("listSourcePeerKeys", AppId);
        }

        public object[] listSourcePeers(ref string AppId)
        {
            return CallMproMFASync("listSourcePeers", AppId);
        }

        public object[] putSourcePeer(ref string AppId, ref string Key, ref string ChStr, ref string SpName, ref string OptStr)
        {
            return CallMproMFASync("putSourcePeer", AppId, Key, ChStr, SpName, OptStr);
        }

        public object[] putWhitelist(ref string AppId, ref string Key, ref string IpAddressStr, ref string OptStr)
        {
            return CallMproMFASync("putWhitelist", AppId, Key, IpAddressStr, OptStr);
        }

        public object[] deleteWhitelist(ref string AppId, ref string Key, ref string IpAddressStr)
        {
            return CallMproMFASync("deleteWhitelist", AppId, Key, IpAddressStr);
        }

        public object[] getSourcePeer(ref string AppId, ref string Key)
        {
            return CallMproMFASync("getSourcePeer", AppId, Key);
        }

        public object[] getWhitelist(ref string AppId, ref string Key)
        {
            return CallMproMFASync("getWhitelist", AppId, Key);
        }

        private object[] CallMproMFASync(string fun, params object[] argsRest)
        {
            return UnwrapResult(CallImemMFASync("mpro_dal_prov", fun, argsRest));
        }
    }
}
